config/master.d/50-master.conf

```diff
@@ -13,6 +13,12 @@ timeout: 20
 file_roots:
   base:
     - /usr/share/salt/kubernetes/salt
+
 pillar_roots:
   base:
     - /usr/share/salt/kubernetes/pillar
+
+ext_pillar:
+  - velum:
+      url: https://localhost/internal-api/v1/pillar.json?version=1
```

Comment:
do we need the `?version=1` param when we have the `v1` namespace already?

Reply:
I think it won't hurt. `v1` is more the version of the whole API in case we introduce not backwards-compatible changes in the future, but `version=1` for the pillar resource (ignored right now) can be really helpful between upgrades, if salt states expect the pillar structure in one way or the other. I'd keep it (despite we are ignoring it in Velum), but I think it can help in case it's needed.

Reply:
No, choose to version one way, not two. Stick to the `/v1/` in the URL. The `/v1/` means your *entire* API is a v1 version, including the data model. @MaximilianMeister is right here.

If you want to change the model in any non-backwards compatible, you *must* rev the API version. We also have to expect that it's possible to have a cluster that needs both v1 and v2 of the API present at one time.

Reply:
In my opinion they are different things. One is the version of the API (`/v1/`). The other is the version of one specific resource within the API (`pillar.json?version=n`), but since it doesn't make sense at the moment I will remove it.

Reply:
This is a few years old now, but it generally lists some good standards for web APIs: https://github.com/WhiteHouse/api-standards

In an API design, you typically version the entire API, not the specific resources in the API. Take a look at the Kubernetes API for an example. They chose to version at the individual resource, which makes sense given the size of their API. My point here is that there aren't multiple levels of versioning in that API.

Reply:
In a public API I'd agree with you without any doubt. However, this API is and was meant to be exclusively used by salt to implement a pillar override (and maybe if decided so, allow it to callback to it and update Velum from salt itself). Branching to new full API versions all endpoints because we decided to refactor the pillar between kubernetes-salt releases would force us to create a new set of controllers for all endpoints, when only the pillar resource contains relevant changes.

I see your point. I completely agree in a general purpose/public API about only having one way of versioning, but I see this API as something a bit different.

In any case I think my argument won't stand because 1) we didn't yet have this problem [why do this? only because we can], and 2) still having two versionings (general api and pillar resource specific) still doesn't sound like a good idea.

Will remove the `version` query parameter.